### PR TITLE
Features/chapter2

### PR DESCRIPTION
**Pull request by voldem007**

### CODE REVIEW

File: SimpleRss/SimpleRss/Models/Feed.swift [hunk 1]
```diff
@@ -8,9 +8,18 @@
 
 import Foundation
 
-struct Feed {
-    let title: String
-    let pubDate: String
-    let picUrl: String
-    let description: String
+class Feed {
+    var isExpanded: Bool
```
Comment (Viktor-by) [inline SimpleRss/SimpleRss/Models/Feed.swift]: Please, avoid adding isExpanded flag in data model. I would consider creating viewModel for cell for such cases.
Look at Feed as a response from server, which must be immutable. We can pass the response to other view controllers that are not aware of purpose of the flag.

**Review by Viktor-by — CHANGES_REQUESTED** [review]

File: SimpleRss/SimpleRss/ViewControllers/FeedViewController.swift [hunk 2]
```diff
@@ -90,7 +93,9 @@ extension FeedViewController: UITableViewDelegate {
         
         guard let cell = tableView.cellForRow(at: indexPath) as? FeedViewCell else { return }
```
Comment (Viktor-by) [inline SimpleRss/SimpleRss/ViewControllers/FeedViewController.swift:91]: let feed = feedList[indexPath.row]

File: SimpleRss/SimpleRss/Services/XMLParserService.swift [hunk 3]
```diff
@@ -8,12 +8,12 @@
 
 import Foundation
 
-class XMLParserService {
-    func fetchXMLData(for url: String?, withCallback completionHandler: @escaping (_ result: [Feed]?, _ error: Error?) -> Void) {
```
Comment (Viktor-by) [inline SimpleRss/SimpleRss/Services/XMLParserService.swift:12]: Although this is not a part of the chapter, but why result cannot be nil?

File: SimpleRss/SimpleRss/ViewControllers/FeedViewController.swift [hunk 4]
```diff
@@ -48,7 +48,7 @@ class FeedViewController: UIViewController {
     func fetchXMLData() {
         XMLParserService().fetchXMLData(for: url) { (feedList, error) in
             if error == nil {
-                self.feedList = feedList!
+                self.feedList = feedList.map { feed in FeedViewModel(feed) }
```
Comment (Viktor-by) [inline SimpleRss/SimpleRss/ViewControllers/FeedViewController.swift]: Why map?

File: SimpleRss/SimpleRss/ViewControllers/FeedViewController.swift [hunk 5]
```diff
@@ -90,7 +93,9 @@ extension FeedViewController: UITableViewDelegate {
         
         guard let cell = tableView.cellForRow(at: indexPath) as? FeedViewCell else { return }
         
-        cell.descriptionLabel.numberOfLines = cell.descriptionLabel.numberOfLines == 0 ? 1 : 0;
+        feedList[indexPath.row].isExpanded = !feedList[indexPath.row].isExpanded
```
Comment (Viktor-by) [inline SimpleRss/SimpleRss/ViewControllers/FeedViewController.swift]: Add a function 'toggle' for FeedViewModel to change its 'isExpanded' flag.
Make setter of isExpanded variable private.

**Review by Viktor-by — CHANGES_REQUESTED** [review]

File: SimpleRss/SimpleRss/Services/XMLParserService.swift [hunk 6]
```diff
@@ -8,13 +8,13 @@
...
-
+class XMLParserService: NSObject {
+    func fetchXMLData(for link: String, withCallback completionHandler: @escaping (_ result: [Feed]) -> Void) {
```
Comment (warokku) [inline SimpleRss/SimpleRss/Services/XMLParserService.swift]: If class doesn't have any properties or internal state, it's probably a good idea to make its funcs `static`

File: SimpleRss/SimpleRss/Views/FeedViewCell.swift [hunk 7]
```diff
@@ -13,8 +13,17 @@ class FeedViewCell: UITableViewCell {
...
     @IBOutlet weak var descriptionLabel: UILabel!
+    @IBOutlet weak var pubDateLabel: UILabel!
     
     override func prepareForReuse() {
         previewImageView.image = nil
+        descriptionLabel.lineBreakMode = .byTruncatingTail
+        descriptionLabel.numberOfLines = 1
     }
+    
+    func expanding(isExpanded: Bool) {
```
Comment (warokku) [inline SimpleRss/SimpleRss/Views/FeedViewCell.swift:29]: Wouldn't it be better to make this a property and put this logic in a setter?

File: SimpleRss/SimpleRss/ViewControllers/FeedViewController.swift [hunk 8]
```diff
@@ -90,7 +89,9 @@ extension FeedViewController: UITableViewDelegate {
         
         guard let cell = tableView.cellForRow(at: indexPath) as? FeedViewCell else { return }
         
-        cell.descriptionLabel.numberOfLines = cell.descriptionLabel.numberOfLines == 0 ? 1 : 0;
+        let feed = feedList[indexPath.row]
+        feed.toggle()
+        cell.expanding(isExpanded: feed.isExpanded)
```
Comment (warokku) [inline SimpleRss/SimpleRss/ViewControllers/FeedViewController.swift]: I believe that cell will be reloaded after calling `tableView.beginUpdates(); tableView. endUpdates()`, so `cell.expanding(isExpanded: feed.isExpanded)` is not really needed here

File: SimpleRss/SimpleRss/Views/FeedViewCell.swift (hunk 7 above)
Comment (voldem007) [reply SimpleRss/SimpleRss/Views/FeedViewCell.swift:29]: Do you mean this approach?
```suggestion
   var isExpanded: Bool {
        get { return expanded }
        set { expanded = newValue }
    }
    
    let feed = feedList[indexPath.row]
    feed.isExpanded != feed.isExpanded
```

File: SimpleRss/SimpleRss/ViewControllers/FeedViewController.swift (hunk 8 above)
Comment (voldem007) [reply SimpleRss/SimpleRss/ViewControllers/FeedViewController.swift]: cell.expanding is existed for change style for label, which allows make description text wider in 
the cell

File: SimpleRss/SimpleRss/Views/FeedViewCell.swift (hunk 7 above)
Comment (warokku) [reply SimpleRss/SimpleRss/Views/FeedViewCell.swift:29]: No, I've meant this:

```suggestion
    var isExpanded: Bool {
        didSet {
            descriptionLabel.numberOfLines = isExpanded ? 0 : 1;
            descriptionLabel.lineBreakMode = isExpanded ? .byCharWrapping : .byTruncatingTail
        }
    }
```

Or you can leave it as it is, but make change to prepareForReuse() code above.

File: SimpleRss/SimpleRss/Views/FeedViewCell.swift [hunk 9]
```diff
@@ -13,8 +13,17 @@ class FeedViewCell: UITableViewCell {
     @IBOutlet weak var previewImageView: UIImageView!
     @IBOutlet weak var titleLabel: UILabel!
     @IBOutlet weak var descriptionLabel: UILabel!
+    @IBOutlet weak var pubDateLabel: UILabel!
     
     override func prepareForReuse() {
```
Comment (warokku) [inline SimpleRss/SimpleRss/Views/FeedViewCell.swift:24]:
```suggestion
    override func prepareForReuse() {
        previewImageView.image = nil
        descriptionLabel.lineBreakMode = .byTruncatingTail
        descriptionLabel.numberOfLines = 1
    }
```
to this:
```suggestion
    override func prepareForReuse() {
        previewImageView.image = nil
        expanding(isExpanded: false)
    }
```

**Pull request merged**